### PR TITLE
add different definitions of permutation

### PR DESCRIPTION
**Pull request by HuStmpHrrr**
implement permutation over lists in two different flavors. these two different flavors are very useful when verifying algorithms over lists. I do know there is another definition lives in `Data.Fin.Permutation`, but I don't think this would work sufficiently handy in some situations.

### CODE REVIEW

**Comment by HuStmpHrrr** [comment]
The proof of `bag-equality⇒perm` is not straightforward. I will need to think more on this part. The inductive case seems to require an inversion lemma that is equivalent to those `drop` lemmas.

**Comment by MatthewDaggitt** [comment]
I very much like the inductive definition in `Data.List.Permutation`. I have a similar one, but mine is much less intuitive to work with. I'd be very happy to accept this in the library. I've added some comments to the code to help bring it slightly more into line with the library style.

I'm not quite so sure about the version in `Data.List.Permutation.DecPropositional`. What is the advantage of this version over the inductive version and bag equality version? Also looking at the code it looks quite difficult to use (e.g. `collect-∷-inv`, `mperm⇒perm` etc.)...

**Comment by JacquesCarette** [comment]
I think it would be good idea to document this as the 'term language' representation of permutations. [There are a plethora of representations for permutations!].  Basically it's a mini programming language that can be interpreted as permutations "when the time comes".  (See my work with Amr Sabry on Pi languages...).

The normal-form problem for such a term language is hard.  This is why the functions to go back-and-forth to other representations are so complicated.  In part because, unlike most other representations, this one isn't unique (i.e. `trans` easily lets you define lots of equivalent permutations that are not propositionally equal).

**Comment by HuStmpHrrr** [comment]
@JacquesCarette thank you for the information. I've never heard of "term language", so I am not sure what's the definition. I perceive all inductive definitions to be a free algebra of some sort. I will definitely document that part along with other changes.

Regarding your work, was "Fractional Types" that you were mentioning?

**Comment by HuStmpHrrr** [comment]
@MatthewDaggitt provided decidable equality, permutation based on counting in fact is the easiest. That's my motivation to have `mperm`. For each element of such type, the problem is effectively reduced to Presburger arithmetic, so really there isn't a need to more complicated APIs and additional lemmas(as they are just lemmas for natural numbers). The inversion lemmas you mentioned are used to prove equivalence to the inductive definition only.

**Comment by HuStmpHrrr** [comment]
I've updated the PR by moving the definitions to the said folder, and also noted it's a term language for permutation.

**Comment by JacquesCarette** [comment]
@HuStmpHrrr Sorry for the slow reply, have been travelling. What I was mentioning was actually the [ESOP 2016 work](https://www.cs.indiana.edu/~sabry/papers/weak-rig-groupoid.pdf) with a whole bunch Agda code [available on github](https://github.com/JacquesCarette/pi-dual) (the important stuff is in the Univalence/ directory).

**Comment by MatthewDaggitt** [comment]
@HuStmpHrrr thanks for changing the directory. A few more minor style tweaks and it's ready to be merged!

**Comment by MatthewDaggitt** [comment]
As for the name of the type, are people generally happy with `Perm`? Anyone prefer infix notation instead? I've been using `_⇿_`, mainly to emphasise that it's a relation, and making names more concise. Not a strong preference though!

**Comment by MatthewDaggitt** [comment]
Oh and also @HuStmpHrrr can you document your changes in the CHANGELOG please?

**Comment by WolframKahl** [comment]
> As for the name of the type, are people generally happy with `Perm`? Anyone prefer infix notation instead? I've been using `_⇿_`, mainly to emphasise that it's a relation, and making names more concise.

I would use `A ⇿ B` for `REL A B`, as in Z.

Z also provides a more complicated arrow for total bijective functions...

**Comment by HuStmpHrrr** [comment]
@MatthewDaggitt i've updated. please take a look.

**Comment by MatthewDaggitt** [comment]
@HuStmpHrrr thanks for the changelog entry. If you look in the "Files changed" tab, and scroll down you should see some comments I've added suggesting a few stylistic changes. Any chance you could address these as well? Sorry I should have pointed out that I'd made these comments earlier!

> I would use `A ⇿ B` for `REL A B`, as in Z.
>
> Z also provides a more complicated arrow for total bijective functions...

@WolframKahl do you have another candidate arrow in mind then? :smile:

**Comment by HuStmpHrrr** [comment]
@MatthewDaggitt I scrolled multiple times but I don't see any comments, nor in my mailbox history. You only commented to move the module, which is shown to be `outdated` before Jacques' first comment. am i missing anything?

**Comment by HuStmpHrrr** [comment]
there is a list of arrows here: http://people.inf.elte.hu/divip/AgdaTutorial/Symbols.html#arrows

> lr~   ↭

this one looks OK to me. the wavy part looks reordering to me, which could serve a good reminder of what this arrow should be.

**Comment by WolframKahl** [comment]
> > Z also provides a more complicated arrow for total bijective functions...
> 
> @WolframKahl do you have another candidate arrow in mind then? :smile: 

Found it:
```
U+2916 	⤖ 	e2 a4 96 	RIGHT­WARDS TWO-HEADED ARROW WITH TAIL
```
:wink:

**Comment by MatthewDaggitt** [comment]
Hmm

> @MatthewDaggitt I scrolled multiple times but I don't see any comments, nor in my mailbox history. You only commented to move the module, which is shown to be outdated before Jacques' first comment. am i missing anything?

Hmm strange. So you don't see something like below when you scroll down to `Data.List.Any.Properties.agda` in the `Files changed tab`?
image

The comments also show up for me in this thread. If you go to the comment `Matthew Daggitt started a review` and click to show all, it should come up with a list of the comments that aren't outdated...

**Comment by HuStmpHrrr** [comment]
@MatthewDaggitt hmm, maybe click the `Finish your review` button? :)

**Comment by MatthewDaggitt** [comment]
> @MatthewDaggitt hmm, maybe click the Finish your review button? :)

Ah thanks! Never formally used the `review` before, previously just submitted comments, and didn't realise I had to click submit. Thanks!

As for arrows, I definitely prefer @HuStmpHrrr 's as it's symmetrical, and I agree the waviness brings to mind re-ordering.

File: src/Data/List/Relation/Permutation/Inductive/Properties.agda [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+  sym-self-inverse (swap x y p)
+    rewrite sym-self-inverse p  = refl
+  sym-self-inverse (trans p₁ p₂)
+    rewrite sym-self-inverse p₁
+          | sym-self-inverse p₂ = refl
+
+  -------------------------------------------------------------------------------------
+  -- _++_
+
+  ++ʳ : ∀ {l₁ l₂ : List A} l → l₁ ⇿ l₂ → l₁ ++ l ⇿ l₂ ++ l
```
Comment (MatthewDaggitt) [inline src/Data/List/Relation/Permutation/Inductive/Properties.agda]: These are introduction forms so they should be called ++⁺ʳ and ++⁺ˡ

File: src/Data/List/Relation/Permutation/Inductive/Properties.agda [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+  ++ʳ l refl         = refl
+  ++ʳ l (prep x p)   = prep x (++ʳ l p)
+  ++ʳ l (swap x y p) = swap x y (++ʳ l p)
+  ++ʳ l (trans p p₁) = trans (++ʳ l p) (++ʳ l p₁)
+
+  ++ˡ : ∀ {l₁ l₂ : List A} l → l₁ ⇿ l₂ → l ++ l₁ ⇿ l ++ l₂
+  ++ˡ [] p      = p
+  ++ˡ (x ∷ l) p = prep x (++ˡ l p)
+
+  ++-cong : ∀ {l₁ l₂ j₁ j₂ : List A} → l₁ ⇿ l₂ → j₁ ⇿ j₂ → l₁ ++ j₁ ⇿ l₂ ++ j₂
```
Comment (MatthewDaggitt) [inline src/Data/List/Relation/Permutation/Inductive/Properties.agda]: This then becomes ++⁺

File: src/Data/List/Relation/Permutation/Inductive/Properties.agda [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+  using (_≡_ ; refl ; _≢_)
+open import Function
+
+open import Data.List.Membership.Propositional.Properties
+open import Data.List.Properties as Listₚ hiding (map-cong)
+
+
+module _ {a} {A : Set a} where
+
+  sym-self-inverse : ∀ {l₁ l₂ : List A} (p : l₁ ⇿ l₂) → sym (sym p) ≡ p
```
Comment (MatthewDaggitt) [inline src/Data/List/Relation/Permutation/Inductive/Properties.agda]: The standard name for this property would be sym-involutive

**Comment by WolframKahl** [comment]
> +open import Relation.Binary.PropositionalEquality as ≡
> 
> We usually use `P` for propositional when qualifying this module instead of `≡`

This has started to change relatively recently,
when people started using such namings for being able to use
```
   ≡.refl
   ≡.trans
   ≡.cong
```
etc.

**Comment by MatthewDaggitt** [comment]
> This has started to change relatively recently,
> when people started using such namings for being able to use
> ```
>    ≡.refl
>    ≡.trans
>    ≡.cong
> ```
> etc.

Hmm that it is very nice, it's true. The rest of the library uses `P` though... Okay, I'm happy to leave it up to people's discretion, although I'd probably prefer to err on the side of consistency.

**Comment by WolframKahl** [comment]
> As for arrows, I definitely prefer @HuStmpHrrr 's as it's symmetrical, and I agree the waviness brings to mind re-ordering.

(But it's use to connect formally potentially different types, so re-ordering does not sound right.)

**Comment by MatthewDaggitt** [comment]
>>As for arrows, I definitely prefer @HuStmpHrrr 's as it's symmetrical, and I agree the waviness brings to mind re-ordering.
>
>(But it's use to connect formally potentially different types, so re-ordering does not sound right.)

I think we might be having a cross purposes conversation? We're trying to come with a symbol for what was permutation, a relation between two elements of type `List A`. Are you discussing `REL A B` by any chance?

**Comment by WolframKahl** [comment]
I was thinking of types `Fin m >->> Fin n` of permutations represented as heterogeneous bijective mappings.
(Where you can infer `m ≡ n`, but in several developments it has turned out to be useful to not assume it.)

**Comment by HuStmpHrrr** [comment]
but really there is no different types here. it's intuitively a symmetric relation between two lists of the same type (and `Data.Fin.Permutation` uses english for that). and i am getting a bit lazier when i have to type a character that is not in the input method.

**Comment by gallais** [comment]
> i am getting a bit lazier when i have to type a character that is not in the input method.

If anything you want is missing from the input method, don't hesitate to open an issue on Agda's
side of things: adding new characters to the input method is straightforward!

**Comment by MatthewDaggitt** [comment]
Thanks for all the updates. Merging it in now. Will make a few more tweaks myself but nothing too huge.

As you've contributed a module @HuStmpHrrr, would you to be added to the list of contributors? If so, under what name?

**Comment by HuStmpHrrr** [comment]
@MatthewDaggitt  thank you. "Jason Hu" is OK. I will try to contribute more regularly :)

**Comment by MatthewDaggitt** [comment]
You've been added to the list.

I've also done a little bit more tidying to bring the module into line with library (see 602b013). I've also generalised some lemmas and added a few more properties. Give me a shout if you think I've accidentally impacted any functionality.

**Pull request merged**